src/reconcile/rec_write.c

```diff
@@ -3729,13 +3753,34 @@ __rec_split_write(WT_SESSION_IMPL *session, WT_RECONCILE *r,
 	/* Write the disk image and get an address. */
 	WT_RET(__wt_bt_write(session,
 	    compressed_image == NULL ? &chunk->image : compressed_image,
-	    addr, &addr_size, false, F_ISSET(r, WT_REC_CHECKPOINT),
-	    compressed_image != NULL));
+	    addr, &addr_size, &compressed_size,
+	    false, F_ISSET(r, WT_REC_CHECKPOINT), compressed_image != NULL));
+	WT_RET(__wt_memdup(session, addr, addr_size, &multi->addr.addr));
+	multi->addr.size = (uint8_t)addr_size;
 #ifdef HAVE_DIAGNOSTIC
 	verify_image = false;
 #endif
-	WT_RET(__wt_memdup(session, addr, addr_size, &multi->addr.addr));
-	multi->addr.size = (uint8_t)addr_size;
+	/*
+	 * Track our compression ratio for this file. We don't want to lock the
+	 * WT_BTREE.compression_adj field, so let it race. For that to work, we
+	 * need a integral value, not a double. Do the compression calculation
+	 * as a double, multiply the result to pick up some decimal places, and
+	 * then truncate it to store it as an integral value. Finally, it's a
+	 * shared memory location and it's not uncommon to be pushing out large
+	 * numbers of pages from the same file, don't bother updating the value
+	 * if there's not at least a 5% shift.
+	 */
+	if (!btree->compressor_is_snappy &&
+	    btree->type != BTREE_COL_FIX && compressed_size) {
+		compression_adj = (uint32_t)(((double)
+		    chunk->image.size / compressed_size) * WT_COMPRESS_ADJ);
+		adjustp = WT_PAGE_IS_INTERNAL(r->page) ?
+		    &btree->intl_compadjust : &btree->leaf_compadjust;
+		if (*adjustp == 0 ||
+		    compression_adj < *adjustp - *adjustp / 20 ||
+		    compression_adj > *adjustp + *adjustp / 20)
```

Comment:
I'd expected this to be more like the algorithm in eviction tuning, that takes historical success into account. i.e something like:

```
#define COMPRESSION_ADJUST_HISTORY_WEIGHT 200
copmadjust = ((compadjust * (COMPRESSION_ADJUST_HISTORY_WEIGHT - 1)) + compression_adj) / COMPRESSION_ADJUST_HISTORY_WEIGHT;
```

Do you think that's not necessary?

Reply:
I thought about considering history when doing this work, but eventually didn't in favor of simplicity. In commit c5a90f4, I made a change so we'll only move 10% (of the shifted value), at a time, that is, outliers will only move the compression ratio by a small amount.